### PR TITLE
Anitha/implement main panel

### PR DESCRIPTION
**Pull request by Anithaus**
## What was changed
add some functions to the index.ts

## Why?
To get started with the main panel, we need feedback to know if we are on the right track.

## Checklist
Used Webview API as a sample to create the sculpt of the function

### CODE REVIEW

File: out/extension.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
+"use strict"
```
Comment (bergundy) [inline out/extension.js:1]: Files in `out` should not be tracked by git.
Let's make sure to put `out` in `.gitignore`.

File: package.json [hunk 2]
```diff
@@ -6,6 +6,18 @@
...
   },
+  "activationEvents": [
+    "onCommand:temporal-debugger-plugin.start"
+  ],
+  "main": "./out/extension.js",
+  "contributes": {
+    "commands": [
+      {
+        "command": "temporal-debugger-plugin.start",
+        "title": "Hello Temporal"
```
Comment (bergundy) [inline package.json:17]: Let's call this `Launch Temporal Debugger Plugin`

File: src/extension_HTML.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+//HTML content of the webview
+export function getWebViewContent(): string {
```
[the remark this answers is not in the document]
Comment (Anithaus) [reply src/extension_HTML.ts]: I changed the extension_HTML.js to main_panel.html, and now it is reading this from disk. Let me know if it works.

File: media/vscode.css [hunk 4]
```diff
@@ -0,0 +1,91 @@
+:root {
```
Comment (bergundy) [inline media/vscode.css:1]: Where did you get these styles?
I'd rather not put too much time into visual design at this point since we might get an actual designer for this plugin eventually.

File: rollup.config.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import commonjs from "@rollup/plugin-commonjs"
+import { terser } from "rollup-plugin-terser"
+import sveltePreprocess from "svelte-preprocess"
+import typescript from "@rollup/plugin-typescript"
+import path from "path"
+import fs from "fs"
+
+const production = !process.env.ROLLUP_WATCH
+
+export default fs.readdirSync(path.join(__dirname, "webviews", "pages")).map((input) => {
```
Comment (bergundy) [inline rollup.config.js:12]: Do we have more than one page?

File: rollup.config.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+      // https://github.com/rollup/plugins/tree/master/packages/commonjs
+      resolve({
+        browser: true,
+        dedupe: ["svelte"],
+      }),
+      commonjs(),
+      typescript({
+        tsconfig: "webviews/tsconfig.json",
+        sourceMap: !production,
+        inlineSources: !production,
```
Comment (bergundy) [inline rollup.config.js:47]: I'd just keep the source maps in a "production" build too.
We're not building for a browser and so the bundle size doesn't affect us too much, better to have the source maps for troubleshooting purposes.

File: src/extension.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
+import * as vscode from "vscode"
+import { StartExtension } from "./startExtension"
+
+// export function activate(context: vscode.ExtensionContext): void {
```
Comment (bergundy) [inline src/extension.ts:4]: Let's avoid committing commented out code

File: src/extension.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+//         path.join(context.extensionPath, "webviews", "components", "main_panel.svelte"),
+//       )
+//       panel.webview.html = fs.readFileSync(filePath.fsPath, "utf8")
+
+//     }),
+//   )
+// }
+
+// this method is called when your extension is activated
+// your extension is activated the very first time the command is executed
```
Comment (bergundy) [inline src/extension.ts:30]: Use jsdoc format to document functions please

File: src/extension.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+export async function activate(context: vscode.ExtensionContext): Promise<any> {
+  console.log('Congratulations, your extension "vstodo" is now active!')
+  await context.subscriptions.push(
+    vscode.commands.registerCommand("temporal-debugger-plugin.start", async () => {
+      await StartExtension.createOrShow(context.extensionUri)
+    }),
+  )
+  //refresing
+  await context.subscriptions.push(
+    vscode.commands.registerCommand("vstodo.refresh", async () => {
```
Comment (bergundy) [inline src/extension.ts:40]: Looks like this was copied as is from some TODO list tutorial, please fix

File: package.json [hunk 10]
```diff
@@ -6,6 +6,23 @@
...
+  "main": "./out/extension.js",
+  "contributes": {
+    "commands": [
+      {
+        "command": "temporal-debugger-plugin.start",
+        "title": "Launch Temporal Debugger Plugin"
+      },
+      {
+        "command": "vstodo.refresh",
+        "title": "Refresh"
```
Comment (bergundy) [inline package.json:22]: Looks like this was copied as is from a todo list tutorial.
Please fix.

File: package.json [hunk 11]
```diff
@@ -17,22 +34,54 @@
...
     "@typescript-eslint/eslint-plugin": "^5.30.0",
     "@typescript-eslint/parser": "^5.30.0",
     "@vscode/test-electron": "^2.1.5",
+    "concurrently": "^7.3.0",
     "eslint": "^8.18.0",
     "eslint-config-prettier": "^8.5.0",
     "eslint-plugin-deprecation": "^1.3.2",
     "eslint-plugin-tsdoc": "^0.2.16",
+    "glob": "^8.0.3",
+    "mocha": "^10.0.0",
```
Comment (bergundy) [inline package.json:57]: Do we need `mocha`? do we run any tests?

File: src/getNonce.ts [hunk 12]
```diff
@@ -0,0 +1,8 @@
+export function getNonce(): string {
```
Comment (bergundy) [inline src/getNonce.ts:1]: Can we use the crypto library for generating this?
Could just use `crypto.randomUUID()`:
https://nodejs.org/docs/latest-v16.x/api/crypto.html#cryptorandomuuidoptions

File: src/startExtension.ts [hunk 13]
```diff
@@ -0,0 +1,145 @@
+import * as vscode from "vscode"
+import { getNonce } from "./getNonce"
+// import * as path from "path"
+// import * as fs from "fs"
```
Comment (bergundy) [inline src/startExtension.ts:4]: Please remove

File: src/startExtension.ts [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+  public static currentPanel: StartExtension | undefined
+
+  public static readonly viewType = "temporal-debugger-plugin"
+
+  private readonly _panel: vscode.WebviewPanel
+  private readonly _extensionUri: vscode.Uri
+  private _disposables: vscode.Disposable[] = []
+
+  public static async createOrShow(extensionUri: vscode.Uri): Promise<any> {
+    const column = vscode.window.activeTextEditor ? vscode.window.activeTextEditor.viewColumn : undefined
```
Comment (bergundy) [inline src/startExtension.ts:19]: Can shorten to:
```suggestion
    const column = vscode.window.activeTextEditor?.viewColumn
```

File: src/startExtension.ts [hunk 15]
```diff
@@ -0,0 +1,145 @@
+import * as vscode from "vscode"
+import { getNonce } from "./getNonce"
+// import * as path from "path"
+// import * as fs from "fs"
+
+export class StartExtension {
```
Comment (bergundy) [inline src/startExtension.ts:6]: Why is this a singleton?
Can't we instantiate it once during startup instead?

File: src/startExtension.ts [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+        await x.dispose()
+      }
+    }
+  }
+
+  private _update(): void {
+    const webview = this._panel.webview
+
+    this._panel.webview.html = this._getHtmlForWebview(webview)
+    webview.onDidReceiveMessage(async (data) => {
```
Comment (bergundy) [inline src/startExtension.ts:88]: This probably shouldn't be in `_update`. If I understand the intent correctly, `_update` might be called multiple times and there's no reason to reregister the event handler when that happens.

File: src/startExtension.ts [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+        </script>
+			</head>
+      <body>    
+			</body>
+			<script src="${scriptUri}" nonce="${nonce}">	</script>
+			</html>`
+  }
+}
+
+// }
```
Comment (bergundy) [inline src/startExtension.ts:145]: Can delete

File: tsconfig.json [hunk 18]
```diff
@@ -1,7 +1,12 @@
 {
   "extends": "@tsconfig/node16/tsconfig.json",
   "compilerOptions": {
+    "lib": ["ES2022"],
```
Comment (bergundy) [inline tsconfig.json:4]: Why did you need to update this? The setting in `@tsconfig/node16` should be good.

File: tsconfig.json [hunk 19]
```diff
@@ -1,7 +1,12 @@
...
     "outDir": "out",
-    "rootDir": "src"
-  }
+    "rootDir": "src",
+    "allowSyntheticDefaultImports": true,
```
Comment (bergundy) [inline tsconfig.json:7]: Is this required?

Comment (bergundy) [reply tsconfig.json:7]: Why?

File: tsconfig.json [hunk 20]
```diff
@@ -1,7 +1,12 @@
...
+    "strict": true,
+    "sourceMap": true
```
Comment (bergundy) [inline tsconfig.json:9]: This is already inherited from the base config.

Comment (bergundy) [reply tsconfig.json:9]: (Except for sourceMap, that can stay)

File: webviews/components/history_panel.html [hunk 21]
```diff
@@ -0,0 +1,25 @@
+<!DOCTYPE html>
```
Comment (bergundy) [inline webviews/components/history_panel.html:1]: Should this be committed here?
Is it just for testing the transition?
Also, this shouldn't be a separate HTML file, just a different view in the same svelte app.

File: webviews/components/main_panel.svelte [hunk 22]
```diff
@@ -0,0 +1,41 @@
+<script>
+  // const vscode = acquireVsCodeApi()
+
+  const handleClick = () => {
+    console.log("Submit successfully!")
+  }
```
Comment (bergundy) [inline webviews/components/main_panel.svelte:6]: Can we make this actually communicate with the vscode side?

File: webviews/components/main_panel.svelte [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+    <input type="checkbox" />
+    <div />
+    <label for="clientCert">Client cert</label>
+    <input type="file" />
+    <div />
+    <label for="clintKey">Client private key</label>
+    <input type="file" />
+  </form>
+  <div id="lines-of-code-counter" />
+  <button on:click={handleClick} id="submit">Submit</button>
```
Comment (bergundy) [inline webviews/components/main_panel.svelte:37]: As mentioned in our sync, let's avoid on:click handlers and put the handler on the form element.

Comment (bergundy) [reply webviews/components/main_panel.svelte:37]: Also don't need the id most likely.

File: .vscode/launch.json [hunk 24]
```diff
@@ -5,6 +5,17 @@
 {
   "version": "0.2.0",
   "configurations": [
+    {
```
Comment (bergundy) [inline .vscode/launch.json:8]: Looks like this was copied from the replayer sample.
Let's keep it for now but it'll probably need to be modified when we get to launching the replayer.

File: package.json [hunk 25]
```diff
@@ -6,6 +6,23 @@
...
+  "contributes": {
+    "commands": [
+      {
+        "command": "temporal-debugger-plugin.start",
+        "title": "Launch Temporal Debugger Plugin"
+      },
+      {
+        "command": "temporal-debugger.history",
+        "title": "Launch Temporal Debugger History"
+      }
```
Comment (bergundy) [inline package.json:23]: I'm assuming this will be removed eventually... we probably only need one command. Am I missing something? Is this used for development?

File: package.json [hunk 26]
```diff
@@ -17,9 +34,16 @@
...
+    "@rollup/plugin-commonjs": "^22.0.1",
+    "@rollup/plugin-node-resolve": "^13.3.0",
+    "@rollup/plugin-typescript": "^8.3.3",
     "@tsconfig/node16": "^1.0.3",
+    "@tsconfig/svelte": "^3.0.0",
     "@types/glob": "^7.2.0",
+    "@types/mocha": "^9.1.1",
     "@types/node": "^16.0.0",
+    "@types/node-fetch": "^2.6.2",
+    "@types/polka": "^0.5.4",
```
Comment (bergundy) [inline package.json:46]: Is this in use?

File: package.json [hunk 27]
```diff
@@ -28,11 +52,38 @@
     "eslint-config-prettier": "^8.5.0",
     "eslint-plugin-deprecation": "^1.3.2",
     "eslint-plugin-tsdoc": "^0.2.16",
+    "glob": "^8.0.3",
```
Comment (bergundy) [inline package.json:55]: Is this in use?

Comment (bergundy) [reply package.json:55]: Let's create an issue to clean up any unused dependencies later instead of tackling this every PR.
There are tools that detect this, but those can be added later too.

File: rollup.config.js [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+          css.write(name + ".css")
+        },
+        preprocess: sveltePreprocess(),
+      }),
+
+      // If you have external dependencies installed from
+      // npm, you'll most likely need these plugins. In
+      // some cases you'll need additional configuration -
+      // consult the documentation for details:
+      // https://github.com/rollup/plugins/tree/master/packages/commonjs
```
Comment (bergundy) [inline rollup.config.js:37]:
```suggestion
```

File: src/extension.ts [hunk 29]
```diff
@@ -0,0 +1,20 @@
...
+export async function activate(context: vscode.ExtensionContext): Promise<void> {
+  context.subscriptions.push(
+    vscode.commands.registerCommand("temporal-debugger-plugin.start", async () => {
+      await StartExtension.createOrShow(context.extensionUri)
+    }),
+  )
+  context.subscriptions.push(
+    vscode.commands.registerCommand("temporal-debugger.history", async () => {
+      await HistoryPanel.createOrShow(context.extensionUri)
+    }),
```
Comment (bergundy) [inline src/extension.ts:18]: Assuming this is just for this development stage and not meant to be used by extension users, if so, please add a comment.

File: src/extension.ts [hunk 30]
```diff
@@ -0,0 +1,20 @@
+import * as vscode from "vscode"
+import { StartExtension } from "./startExtension"
+import { HistoryPanel } from "./histryExtension"
+
+/*
+ * This method is called when your extension is activated
+ * Your extension is activated the very first time the command is executed
+ * */
```
Comment (bergundy) [inline src/extension.ts:8]: Don't need to leave the tutorial comments
```suggestion
```

File: src/getNonce.ts [hunk 31]
```diff
@@ -0,0 +1,7 @@
+import crypto from "crypto"
```
Comment (bergundy) [inline src/getNonce.ts:1]: I think this file isn't used anymore..

Comment (bergundy) [reply src/getNonce.ts:1]: I'm wrong, maybe I saw it removed in a different version of the code during our zoom meeting.

File: src/histryExtension.ts [hunk 32]
```diff
@@ -0,0 +1,114 @@
+import * as vscode from "vscode"
+
+import { getNonce } from "./getNonce"
+
+export class HistoryPanel {
```
Comment (bergundy) [inline src/histryExtension.ts:5]: I have some ideas how to make this less static but still use the singleton pattern.

File: src/replay_history/client.ts [hunk 33]
```diff
@@ -0,0 +1,22 @@
+import { WorkflowClient } from "@temporalio/client"
```
Comment (bergundy) [inline src/replay_history/client.ts:1]: Let's put all of this in a `samples` directory.

File: src/startExtension.ts (hunk 15 above)
Comment (bergundy) [reply src/startExtension.ts:5]: Same here, let's discuss alternative patterns.

File: src/histryExtension.ts [hunk 34]
```diff
@@ -0,0 +1,114 @@
+import * as vscode from "vscode"
```
Comment (bergundy) [inline src/histryExtension.ts:1]: This is mostly duplicate of `startExtension`, we should refactor this into one class.

File: tsconfig.json [hunk 35]
```diff
@@ -3,5 +3,6 @@
   "compilerOptions": {
     "outDir": "out",
     "rootDir": "src"
-  }
+  },
+  "exclude": ["node_modules", "webviews"]
```
Comment (bergundy) [inline tsconfig.json:7]: Why is this necessary?

File: webviews/components/main_panel.svelte [hunk 36]
```diff
@@ -0,0 +1,67 @@
...
+
+    for (let field of formData) {
+      const [key, value] = field
+      data[key] = value
+    }
+
+    vscode.postMessage({
+      type: "onSubmit",
+    })
+    console.log(data)
```
Comment (bergundy) [inline webviews/components/main_panel.svelte:34]: I'd add some comments here explaining that this is work in progress and what it should eventually do.

File: webviews/components/main_panel.svelte [hunk 37]
```diff
@@ -0,0 +1,67 @@
...
+    <input type="submit" value="Start" />
+  </form>
+  <hr />
+  <p>Debug from history file</p>
+  <form on:submit|preventDefault={historyFile}>
+    <input type="file" required name="File" />
+    <input type="submit" value="Start" />
+  </form>
+  <hr />
+  <p>Configure server credentials (for downloading histories)</p>
```
Comment (bergundy) [inline webviews/components/main_panel.svelte:52]: Eventually this will probably be a different view, not critical for now though.

File: webviews/pages/history_panel.ts [hunk 38]
```diff
@@ -0,0 +1,7 @@
+import App from "../components/history_panel.svelte"
```
Comment (bergundy) [inline webviews/pages/history_panel.ts:1]: Not needed as discussed before.

File: rollup.config.js [hunk 39]
```diff
@@ -43,6 +38,7 @@ export default fs.readdirSync(path.join(__dirname, "webviews", "pages")).map((in
       typescript({
         tsconfig: "webviews/tsconfig.json",
       }),
+      json(),
```
Comment (bergundy) [inline rollup.config.js:41]: You're not going to need this

File: src/extension.ts [hunk 40]
```diff
@@ -1,20 +1,12 @@
 import * as vscode from "vscode"
 import { StartExtension } from "./startExtension"
-import { HistoryPanel } from "./histryExtension"
 
-/*
- * This method is called when your extension is activated
- * Your extension is activated the very first time the command is executed
- * */
+//extension is activated
```
Comment (bergundy) [inline src/extension.ts:4]: I'd just remove this comment, it's not helpful

File: webviews/components/history_panel.svelte [hunk 41]
```diff
@@ -1,4 +1,4 @@
-<script lang="ts">
+<script leng="ts">
```
Comment (bergundy) [inline webviews/components/history_panel.svelte:1]: typo

File: webviews/tsconfig.json [hunk 42]
```diff
@@ -2,5 +2,5 @@
   "extends": "@tsconfig/svelte/tsconfig.json",
   "include": ["./**/*"],
   "exclude": ["../node_modules/*"],
-  "compilerOptions": { "types": ["svelte"], "strict": true }
+  "compilerOptions": { "types": ["svelte"], "strict": true, "moduleResolution": "node", "resolveJsonModule": true }
```
Comment (bergundy) [inline webviews/tsconfig.json:5]: You don't need all of this

File: webviews/globals.d.ts [hunk 43]
```diff
@@ -0,0 +1,6 @@
+import * as vscode from "vscode"
+
+declare global {
+  const tsvscode: any
+  const data: any
```
Comment (bergundy) [inline webviews/globals.d.ts:5]: Not sure why these globals are needed

File: webviews/components/main_panel.svelte [hunk 44]
```diff
@@ -58,10 +70,14 @@
...
     <label for="clintKey">Client private key</label>
-    <input type="file" name="Client private key" />
+    <input type="file" name="Client_private_key" />
     <button type="submit">Submit</button>
   </form>
 </main>
+<!-- showing history component on click of submit button-->
+{#if views}
+  <svelte:component this={HistoryView} />
+{/if}
```
Comment (bergundy) [inline webviews/components/main_panel.svelte:83]: This should be in a separate view, not under the main panel
